core/testsuite/src/test/java/com/blazebit/persistence/testsuite/OptimizedKeysetPaginationRowValueConstructorTest.java

```diff
@@ -0,0 +1,414 @@
+                Person o3 = new Person("Karl3");
+                o4 = new Person("Karl4");
+
+                doc1.setOwner(o1);
+                doc2.setOwner(o2);
+                doc3.setOwner(o3);
+                doc4.setOwner(o4);
+
+                em.persist(o1);
+                em.persist(o2);
+                em.persist(o3);
+                em.persist(o4);
+
+                em.persist(doc1);
+                em.persist(doc2);
+                em.persist(doc3);
+                em.persist(doc4);
+            }
+        });
+    }
+
+    @Override
+    protected CriteriaBuilderConfiguration configure(CriteriaBuilderConfiguration config) {
+        config = super.configure(config);
+        config.setProperty(ConfigurationProperties.OPTIMIZED_KEYSET_PREDICATE_RENDERING, "true");
+        return config;
+    }
+
+    @Test
+    public void simpleNormalTest() {
+        CriteriaBuilder<Tuple> crit = cbf.create(em, Tuple.class).from(Document.class, "d")
+                .select("d.name").select("d.owner.name");
+        crit.orderByDesc("d.owner.name")
+                .orderByAsc("d.name")
+                .orderByAsc("d.id");
+
+        PaginatedCriteriaBuilder<Tuple> pcb = crit.page(null, 1, 1);
+        PagedList<Tuple> result = pcb.getResultList();
+        simpleTest(crit, pcb, result);
+    }
+
+    @Test
+    public void backwardsPaginationResultSetOrder() {
+        CriteriaBuilder<Tuple> crit = cbf.create(em, Tuple.class).from(Document.class, "d")
+                .select("d.name").select("d.owner.name");
+        crit.orderByDesc("d.owner.name")
+                .orderByDesc("d.name")
+                .orderByAsc("d.id");
+
+        PaginatedCriteriaBuilder<Tuple> pcb = crit.page(null, 2, 1);
+        PagedList<Tuple> result = pcb.getResultList();
+
+        // scroll backwards
+        result = crit.page(result.getKeysetPage(), 0, 2).getResultList();
+
+        assertEquals(2, result.getSize());
+        assertEquals("doc4", result.get(0).get(0));
+        assertEquals("doc3", result.get(1).get(0));
+    }
+
+    @Test
+    public void backwardsPaginationWithCollectionResultSetOrder() {
+        CriteriaBuilder<Tuple> crit = cbf.create(em, Tuple.class).from(Document.class, "d")
+                .select("d.name").select("d.owner.name").select("d.people");
+        crit.orderByDesc("d.owner.name")
+                .orderByDesc("d.name")
+                .orderByAsc("d.id");
+
+        PaginatedCriteriaBuilder<Tuple> pcb = crit.page(null, 2, 1);
+        PagedList<Tuple> result = pcb.getResultList();
+
+        // scroll backwards
+        result = crit.page(result.getKeysetPage(), 0, 2).getResultList();
+
+        assertEquals(2, result.getSize());
+        assertEquals("doc4", result.get(0).get(0));
+        assertEquals("doc3", result.get(1).get(0));
+    }
+
+    @Test
+    public void forwardBackwardsPaginationResultSetOrder() {
+        final Document doc5 = new Document("doc5");
```

Comment:
I suppose the MySQL issues(https://travis-ci.org/Blazebit/blaze-persistence/jobs/266267687) might be related to the fact that you are using `transactional` to persist new entities after you started the TX for your test. To fix this I recommend following the suggestion from above i.e. using the same data set throughout a single test class.